src/common/transformations/src/transformations/common_optimizations/align_mixed_fp32_fp16_types.cpp

```diff
@@ -0,0 +1,89 @@
+            const auto& incoming_output = input.get_source_output();
+            const auto& incoming_node = incoming_output.get_node_shared_ptr();
+
+            if (fp16_compression_is_disabled(incoming_node))
+                continue;  // we are in the middle
+
+            if (!incoming_output.get_element_type().is_real())
+                continue;
+
+            const auto& existing_convert = std::dynamic_pointer_cast<opset10::Convert>(incoming_node);
+            if (existing_convert) {
+                existing_convert->set_convert_element_type(element::f32);
+            } else {
+                auto convert = std::make_shared<opset10::Convert>(incoming_output, ov::element::f32);
+                convert->set_friendly_name(incoming_node->get_friendly_name() + "_decompressed_to_f32");
+                copy_runtime_info(incoming_node, convert);
+                input.replace_source_output(convert);
+                ov::disable_fp16_compression(convert);
+            }
+            is_changed = true;
+        }
+        return is_changed;
+    };
+
+    std::function<bool(const std::shared_ptr<Node>&)> insert_converts_after = [&](const std::shared_ptr<Node>& node) {
+        bool is_changed = false;
+        for (const auto& output : node->outputs()) {
+            for (const auto& out_inputs : output.get_target_inputs()) {
+                auto out_node = out_inputs.get_node()->shared_from_this();
+                if (fp16_compression_is_disabled(out_node) || is_precision_sensitive(out_inputs))
+                    continue;
+                if (!out_inputs.get_element_type().is_real())
+                    continue;
+
+                // todo xxx-101766: if we don't skip Results there is an error on GPU
+                if (std::dynamic_pointer_cast<opset10::Result>(out_node))
+                    continue;
+
+                const auto& existing_convert = std::dynamic_pointer_cast<opset10::Convert>(out_node);
+
+                if (existing_convert) {
+                    existing_convert->set_convert_element_type(element::f32);
+                } else {
+                    // element_type of this convert will be changed automatically to f16 after
+                    // ConvertPrecision(f32 -> f16). It's kept here f32 to keep ov::Model validatable
+                    auto convert = std::make_shared<opset10::Convert>(output, element::f32);
+                    copy_runtime_info(node, convert);
+                    convert->set_friendly_name(node->get_friendly_name() + "_compressed_to_f16");
+                    out_inputs.replace_source_output(convert);
+                }
+                is_changed = true;
+            }
+        }
+        return is_changed;
+    };
+
+    bool is_changed = false;
+    for (auto& node : model->get_ordered_ops()) {
+        if (!fp16_compression_is_disabled(node))
+            continue;
+
+        is_changed |= insert_converts_before(node);
+        is_changed |= insert_converts_after(node);
```

Comment:
We must insert Converts to `f32` before we enter precision sensitive subgraph (if it's needed) and Converts to `f16` after we exit precision sensitive subgraph.

If we have for example MVN node (which is precision sensitive) there will be one Convert at before MVN to upcast to f32 and one after to downcast to f16 both of the functions will do their part and both these converts are needed. 

If there are for example MVN -> some reshape, StridedSlice -> ReduceSum -> and again MVN then Converts will be inserted before the first MVN and after the last one but not between them. Whether we need or not to insert is decided inside functions them-self. 

To clarify that these converts are inserted only if they are needed i renamed them into `insert_converts_after_if_needed`, `insert_converts_before_if_needed`.

About revalidation, it makes to sense to do it here. Indeed the whole model here still remains in `f32` we do not change any node types here, we insert necessary `Converts` so that after `ConvertPrecision to f16` (i) only necessary nodes will be downcasted to f16 and (ii) no type inconsistency will happen on Elementwise, Concat, etc.